compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -215,22 +212,19 @@ impl<'tcx> Inliner<'tcx> {
         }
 
         if let Some(callee_def_id) = callee.def_id().as_local() {
-            let callee_hir_id = self.tcx.hir().local_def_id_to_hir_id(callee_def_id);
             // Avoid a cycle here by only using `instance_mir` only if we have
-            // a lower `HirId` than the callee. This ensures that the callee will
-            // not inline us. This trick only works without incremental compilation.
-            // So don't do it if that is enabled.
-            if !self.tcx.dep_graph.is_fully_enabled() && self.hir_id.index() < callee_hir_id.index()
+            // a lower `DefPathHash` than the callee. This ensures that the callee will
+            // not inline us. This trick even works with incremental compilation,
+            // since `DefPathHash` is stable.
+            if self.tcx.def_path_hash(caller_def_id).local_hash()
+                < self.tcx.def_path_hash(callee_def_id.to_def_id()).local_hash()
```

Comment:
Maybe use `callee.def_id()` here instead of calling `.to_def_id()` after turning it into a local defid? Then the `if let` above can become `if ... .is_some()`.